### ISSUE TITLE
No linear space operators defined on groups

### ISSUE THREAD

**Issue by EelcoHoogendoorn** [brentyi/jaxlie#2]
I noticed that the objects in jaxlie do not define addition or scalar multiplication on elements in the group. Is this simply not implemented yet, or am I missing a good reason why youd want to omit such functionality out of the box?

Somewhat related to that; if I define for instance a class Quaternion(SO3), where I add those linear operators myself, I run into some trouble with the design. Now multiplying Quats will return an SO3, etc. I suppose one way of mitigating that would be to use subclass-friendly type(self) style constructors in the base classes; but again there may be reasons why this is not desirable?

Thinking about further generalizations; it seems to me that this programming style in JAX goes well together with implementing a more general geometric-algebra package, like [clifford](https://clifford.readthedocs.io/en/latest/index.html). Their emphasis is on expressivity rather than performance but I think given JAX compilation you dont really have to choose and youll get all the performance that JAX can offer at least. But yeah, that would be a bit of a project.

**Comment by brentyi** [comment]
Hey, sorry for missing this! 

Can you clarify (or link some references on) the expected behavior of the linear space operators that you mentioned? This isn't ringing any bells and I may be missing something obvious.

Subclass-friendly constructors: this would chip away at readability, but that's the only reason I can think of to not incorporate them. I wouldn't be opposed if there are good use cases. :slightly_smiling_face: 

General programming style: agree that a more general package would be cool! I've been pleasantly surprised with how well the whole everything-is-a-dataclass approach has turned out, though it does start feeling hacky once function transformations (vmap, grad, etc) start getting incorporated.

**Comment by EelcoHoogendoorn** [comment]
https://en.wikipedia.org/wiki/Vector_space

Basically I just mean addition of two objects of equal type, and scalar multiplication; __add__(self, other), and __mul__(self, scalar). These should be well defined operations for all the classes considered in this package I think? I quickly run into a need for those if implementing any algorithms using complex numbers or quaternions; and I feel theyd be better off as part of the package.

**Comment by brentyi** [comment]
So behavior like this?
```python
from typing import TypeVar
import jaxlie
import jax

T = TypeVar("T", bound=jaxlie.MatrixLieGroup)

def add(x: T, y: T) -> T:
     return type(x)(x.parameters() + y.parameters())
     # alternatively: return jax.tree_map(lambda a, b: a + b, x, y)

def multiply(scalar: float, x: T) -> T:
     return type(transform)(scalar * transform.parameters())
     # alternatively: return jax.tree_map(lambda a: scalar * a, x)
```

The reason these aren't present is that the provided classes are intended to be interfaces for matrix Lie groups, and not general complex numbers or quaternions. SO(3), for example, provides an API for 3x3 special (positive determinant) orthogonal matrices, which (a) happen to be compactly parameterized by quaternions, and (b) live on a manifold that's closed under neither addition nor scalar multiplication. We've also been assuming a unit norm constraint on complex numbers and quaternions, which would break if we performed linear addition/multiplication operations on the embedding space.

Does that make sense?

On vmap: glad you figured that out! Once things click it's definitely really convenient. Might add something to the README...

**Comment by EelcoHoogendoorn** [comment]
Right; it makes sense that the objects as named are intended to be normalized quantities; so indeed it makes sense not to provide operations that take them outside their designated type. Still the quaternion based algorithm im implementing has a step of the form (quat + delta).normalize(), so while perhaps confusing from a math-purity point of view, you could make a pragmatic case for it.

That being said, perhaps what that really means is that I should be creating my own jax-geometric-algebra package, rather than trying to shoehorn a lie-group focused one into something it isnt.

**Comment by brentyi** [comment]
Are you able to share the algorithm you're implementing? Mostly wondering how frequently an operation like this is actually needed. If it's not often, writing out `SO3(quat.wxyz + delta.wxyz).normalize()` explicitly feels a little clearer to me.

Do let me know if you make a more general package though, sounds like something that could be useful to me. :)

**Comment by EelcoHoogendoorn** [comment]
> Do let me know if you make a more general package though, sounds like something that could be useful to me. :)

https://github.com/EelcoHoogendoorn/numga

There you go!
